nntrainer/include/tensor.h

```diff
@@ -41,6 +41,7 @@ extern "C" {
 
 namespace nntrainer {
 
+class DeferredTensor;
```

Comment:
how about implement tensor directly. Cause if this works well, we are not going to use tensor any way. Isn't it? and also it looks like duplication to have add and add_i

Reply:
1. About deprecating existing operator (change `add_i` to `add`):

I think this could be done after we change all traditional operation used in the code base and be sure that the traditional version is not needed in all case since mechanism is slightly different. 

for example:
```cpp
 Tensor t = t.add(3.14);
``` 
won't work if we change now. 

2. Combining this to tensor:

IMHO it is better to leave monadic wrapper to have a separate class handling lazy evaluation if we are going to use the same method name. 

If we migrate this to tensor, we need 1. a status added to tell if it is lazy evaluating and 2. case handling for every operation and 3. same code bits in every operation methods.

more specifically this could be done like this but IMO less practical:
```cpp
Tensor Tensor::some_operation(float const &value) {
 /* code bits that should be added for every operation */
  if(this->chained) {
    call_chain.push_back([value](Tensor &t) mutable -> int {
      this->some_operation(value);
      return ML_ERROR_NONE;
    });
    return *this;
  }
 /* handle real operations */
}
```

As for as I can tell, I think it is better to leave as it is for now.
Please tell me if those can be implemented in better way!

Reply:
```Tensor t = t.add(3.14); ``` I think we are not going to use this after this deferred tensor is enabled. May be use this ``` Tensor t = t.chain().add(3.14).run()```. And when we do the chain(), it copy the tensor. I think it looks like tensor has tensor copy and do the function which is added by the numeric function ( current functions like add() is not going to use. and if we update add_i to add, then it add function into call_chain. and run() and save the data into copied tensor and may be update the original if there is assign operation. How about it?

Reply:
We had agreement after some valuable discussions. Yes. we are going to use this approach. But change Class name with LazyTensor to make more clear. Thanks.

Reply:
Thanks for the suggestion :) I'll change the class name and rebase other PR 👍